### DISCUSSION TITLE
Canonicalize valid, canonicalized SMILES raises ArgumenError exception

### DISCUSSION THREAD

**Discussion by hogru** [rdkit/rdkit#5443] [Q&A] [answered]
I am not a heavy user of `rdkit` so it might well be that this is a based on wrong assumptions: I generate molecules and try to assess whether a molecule is valid. If it is, I canonicalize it. The issue is, that the generated molecules are checked/validated again later in the pipeline and some of those molecules seem not to be valid (can not be canonicalized)

This is a (simplified) example source code for illustration.

```py
from rdkit import Chem

smiles = "[O-]c1ccc(C[C@H]2NC(c3ccc(Cl)cc3)=n2c2nc(-c3ccco3)cs2)cc1"
print(f"The original SMILES: {smiles}")

print("Check if this is a valid SMILES...", end="")
mol = Chem.MolFromSmiles(smiles)

if mol is None:
    print("Invalid SMILES, stop here")
else:
    print("It IS valid, canonicalize...")
    canon_smiles = Chem.CanonSmiles(smiles)
    print(f"\nCanonicalized SMILES: {canon_smiles}")  # the first n became an N

    print("\nNow, canonicalize the canonicalized SMILES again...")
    print("Expectation: works, no change in canonical SMILES; in reality: raises ArgumentError exception")
    canon_smiles = Chem.CanonSmiles(canon_smiles)  # throws an exception (could handle it, but why?)
    print(f"\nCanonicalized SMILES: {canon_smiles}")
```

**Comment by ptosco** [comment, marked as answer]
@hogru Your initial SMILES is quite peculiar as it specifies the 4-membered ring as part aliphatic (first 3 atoms) and part aromatic (last nitrogen). This seems to confuse the kekulizer causing it to generate an invalid SMILES that it cannot be parsed anymore.

**Reply by hogru** [reply]
@ptosco Thank you very much for your quick response! I understand that this specific molecule is rather strange and I can‘t argue with that. But it seems that ˋrdkitˋ could handle this more gracefully, agreed? Therefore, I might report this as a bug since a molecule generator might come up with all sorts of weird molecules and I need to check their validity somehow (again, I might miss more preferred ways to do that)
